### ISSUE TITLE
TH1F bin content saturation

### ISSUE THREAD

**Issue by stwunsch** [root-project/root#6671]
See for comparison the screenshot below.

The upper plot was done with `TTree.Draw`:

```python
>>> import ROOT
>>> f = ROOT.TFile('DYJetsToLL.root')
>>> t = f.Get('Events')
>>> t.Draw('GenPart_pdgId')
```

The lower plot was done with `RDataFrame.Histo1D`:

```python
>>> import ROOT
>>> c = ROOT.TCanvas()
>>> h = ROOT.RDataFrame('Events', 'DYJetsToLL.root').Histo1D('GenPart_pdgId')
>>> h.Draw()
```

![Screenshot from 2020-10-20 10-38-16](https://user-images.githubusercontent.com/6951222/96562487-1974b880-12c1-11eb-9bc5-1006c78031e5.png)

I've used ROOT 6.22/02 and you can download the file here:

http://opendata.web.cern.ch/record/12353

**Comment by bellenot** [comment]
@stwunsch which one is correct?

**Comment by stwunsch** [comment]
Good question! I guess the lower one, also checked by counting only the respective entries in the `GenPart_pdgId` vector.

**Comment by stwunsch** [comment]
Well, it's a DY sample, so physically spoken we would expect the upper one (?) Let's say, it's not so clear what's going on :sweat_smile:

**Comment by eguiraud** [comment]
Can you enforce the same exact binning in both and also try with raw TTree and raw TTreeReader? If RDF or TTree is producing wrong histograms it's a critical issue if not a release blocker.

**Comment by stwunsch** [comment]
@eguiraud Do we agree that the code above should produce (in principle) the same output?

**Comment by stwunsch** [comment]
@eguiraud Here you can find the plots produced with the same binning.

```python
>>> import ROOT
>>>  f = ROOT.TFile('DYJetsToLL.root')
>>> t = f.Get('Events')
>>> h = ROOT.TH1F('h', 'h', 100, -20, 20)
>>> c = ROOT.TCanvas()
>>> t.Draw('GenPart_pdgId>>h')
```

```python
>>> import ROOT
>>> c = ROOT.TCanvas()
>>> h = ROOT.RDataFrame('Events', 'DYJetsToLL.root').Histo1D(('h', 'h', 100, -20, 20), 'GenPart_pdgId')
>>> h.Draw()
```

![Screenshot from 2020-10-20 11-33-43](https://user-images.githubusercontent.com/6951222/96568572-63ad6800-12c8-11eb-8488-dcd108b019a8.png)

**Comment by stwunsch** [comment]
As a cross check: The manual count in RDF of the entries we see in the histogram.

```python
>>> ROOT.RDataFrame('Events', 'DYJetsToLL.root').Define('n', 'Sum(GenPart_pdgId == 11)').Sum('n').GetValue()
36148151.0
>>> ROOT.RDataFrame('Events', 'DYJetsToLL.root').Define('n', 'Sum(GenPart_pdgId == 13)').Sum('n').GetValue()
25642431.0
>>> ROOT.RDataFrame('Events', 'DYJetsToLL.root').Define('n', 'Sum(GenPart_pdgId == 15)').Sum('n').GetValue()
18289536.0
```

**Comment by eguiraud** [comment]
> Do we agree that the code above should produce (in principle) the same output?

I don't see why not. We really need an authoritative answer as to which of the plots is correct though. Maybe raw TTree and TTreeReader can act as tie-breaker. There are no missing entries (the number of entries in the two histograms is the same) so it seems that one of the two interfaces is reading wrong values -- but how can it be that histogram mean and variance are exactly the same in the two cases? The mean, I can understand, but the variance should be larger for the upper histo than for the lower histo.

**Comment by stwunsch** [comment]
Here a reproducer with my best raw TTree skills:

```cpp
void test() {
    auto f = TFile::Open("DYJetsToLL.root");
    auto t = f->Get<TTree>("Events");
    UInt_t n;
    Int_t x[100];
    t->SetBranchAddress("nGenPart", &n);
    t->SetBranchAddress("GenPart_pdgId", x);
    auto n11 = 0u;
    auto n13 = 0u;
    auto n15 = 0u;
    for(auto i = 0u; i < t->GetEntries(); i++) {
        t->GetEntry(i);
        for (auto j = 0u; j < n; j++) {
            if (x[j] == 11) n11++;
            if (x[j] == 13) n13++;
            if (x[j] == 15) n15++;
        }
    }
    cout << "n11: " << n11 << endl;
    cout << "n13: " << n13 << endl;
    cout << "n15: " << n15 << endl;
}
```

```
n11: 36148151
n13: 25642431
n15: 18289536
```

So we can conclude that's a plotting issue in `TTree::Draw`?

**Comment by stwunsch** [comment]
Another ingredient, `tree.Draw('GenPart_pdgId[0]')` gives a correct plot:
![Screenshot from 2020-10-20 12-25-09](https://user-images.githubusercontent.com/6951222/96574248-7e370f80-12cf-11eb-97c6-8bc8f212913b.png)

**Comment by eguiraud** [comment]
Ok, thanks for investigating. I think it's safe to treat it as a `TTree::Draw` bug unless @pcanal figures out otherwise.

**Comment by eguiraud** [comment]
Ah, one last thing: could it be there are _that many_ points that are right at the bin edges? it would explain why mean and variance are the same in the two cases. Can you try with a less regular or much more fine-grained binning?

**Comment by stwunsch** [comment]
If the edge would be the issue, we would see more bins filled than the six we have in the plot, right?

**Comment by stwunsch** [comment]
This should be the used bins:

```python
>>> import numpy as np
>>> np.linspace(-20, 20, 101)
array([-20. , -19.6, -19.2, -18.8, -18.4, -18. , -17.6, -17.2, -16.8,
       -16.4, -16. , -15.6, -15.2, -14.8, -14.4, -14. , -13.6, -13.2,
       -12.8, -12.4, -12. , -11.6, -11.2, -10.8, -10.4, -10. ,  -9.6,
        -9.2,  -8.8,  -8.4,  -8. ,  -7.6,  -7.2,  -6.8,  -6.4,  -6. ,
        -5.6,  -5.2,  -4.8,  -4.4,  -4. ,  -3.6,  -3.2,  -2.8,  -2.4,
        -2. ,  -1.6,  -1.2,  -0.8,  -0.4,   0. ,   0.4,   0.8,   1.2,
         1.6,   2. ,   2.4,   2.8,   3.2,   3.6,   4. ,   4.4,   4.8,
         5.2,   5.6,   6. ,   6.4,   6.8,   7.2,   7.6,   8. ,   8.4,
         8.8,   9.2,   9.6,  10. ,  10.4,  10.8,  11.2,  11.6,  12. ,
        12.4,  12.8,  13.2,  13.6,  14. ,  14.4,  14.8,  15.2,  15.6,
        16. ,  16.4,  16.8,  17.2,  17.6,  18. ,  18.4,  18.8,  19.2,
        19.6,  20. ])
```

The values -15, -13, -11, 11, 13 and 15 are not on the edges.

**Comment by eguiraud** [comment]
Fair enough :smile:

**Comment by eguiraud** [comment]
An interesting observation: if you look at the bars in the two plots, for the RDF plot you can estimate the total number of entries as `37e6*2 + 26e6*2 + 18e6*2 = 1.62e8` which is consistent with the number of entries reported by the histogram. For the `TTree::Draw` plot you get, more or less, `16.75e6 * 6 = 1.01e8`: why are the other entries not displayed? Did they end up in the overflow/underflow bins?

**Comment by stwunsch** [comment]
No underflows or overflows:

```python
>>> import ROOT
>>> f = ROOT.TFile('DYJetsToLL.root')
>>> t = f.Get('Events')
>>> c = ROOT.TCanvas()
>>> h = ROOT.TH1F('h', 'h', 100, -20, 20)
>>> t.Draw('GenPart_pdgId>>h')
>>> h.GetBinContent(0)
0.0
>>> h.GetBinContent(21)
0.0
```

**Comment by stwunsch** [comment]
Last round of investigations, here the `Events->Print()` output:

```
*............................................................................*
*Br   62 :nGenPart  : nGenPart/i                                             *
*Entries : 30458871 : Total  Size=  121869654 bytes  File Size  =   27764921 *
*Baskets :      349 : Basket Size=     574464 bytes  Compression=   4.39     *
*............................................................................*
*Br   67 :GenPart_pdgId : GenPart_pdgId[nGenPart]/I                          *
*Entries : 30458871 : Total  Size=  767472277 bytes  File Size  =  175369010 *
*Baskets :      698 : Basket Size=    1738752 bytes  Compression=   4.38     *
*............................................................................*
```

And the issue is also visible in `TBrowser` by clicking on the branch:

![Uploading Screenshot from 2020-10-20 13-26-28.png…]()

**Comment by eguiraud** [comment]
![image](https://user-images.githubusercontent.com/10999034/96580341-6fa12600-12d8-11eb-92e2-7fd278c6fe0d.png)

**Comment by pcanal** [comment]
Running TTree::Scan should give an idea of the type of error made by TTree::Draw.

**Comment by stwunsch** [comment]
`TTree::Scan` looks fine, I guess.

```
$ root -l DYJetsToLL.root 
root [0] 
Attaching file DYJetsToLL.root as _file0...
(TFile *) 0x55a4bb156060
root [1] Events->Scan("GenPart_pdgId")
***********************************
*    Row   * Instance * GenPart_p *
***********************************
*        0 *        0 *        15 *
*        0 *        1 *        11 *
*        0 *        2 *       -15 *
*        0 *        3 *        11 *
*        0 *        4 *       -11 *
*        0 *        5 *        11 *
*        0 *        6 *        15 *
*        0 *        7 *       -11 *
*        0 *        8 *        15 *
*        0 *        9 *        11 *
*        0 *       10 *        15 *
*        0 *       11 *       -15 *
*        0 *       12 *        15 *
*        0 *       13 *        15 *
*        0 *       14 *        11 *
*        0 *       15 *        11 *
*        0 *       16 *        15 *
*        0 *       17 *        15 *
*        0 *       18 *        11 *
*        1 *        0 *       -13 *
*        1 *        1 *       -13 *
*        1 *        2 *       -11 *
*        1 *        3 *        13 *
*        1 *        4 *        11 *
*        1 *        5 *        13 *
```

**Comment by eguiraud** [comment]
@pcanal ping (this is marked as critical)

**Comment by eguiraud** [comment]
Found a smoking gun with @stwunsch : given a file with more than 16777216 entries, let's say, 50e6:

```python
>>> ROOT.RDataFrame(50000000).Define("x", "1").Snapshot("t", "f.root")
```

`TTree::Draw` caps the contents of each bin at 16777216 (note in particular that `h->GetEntries()` returns the correct number, 50e6, but the bin contents are wrong:

```cpp
root [1] t->Draw("x>>h")
Info in <TCanvas::MakeDefCanvas>:  created default TCanvas with name c1
root [2] h->GetEntries()
(double) 50000000.
root [3] h->GetBinContent(2)
(double) 16777216.
root [4] h->GetBinContent(0)
(double) 0.0000000
root [5] h->GetBinContent(1)
(double) 0.0000000
root [6] h->GetBinContent(3)
(double) 0.0000000
```

@pcanal help

**Comment by pcanal** [comment]
So, if you run on a few entries, the result are consistent between RDataFrame and TTree::Draw?

**Comment by eguiraud** [comment]
Yes

**Comment by eguiraud** [comment]
Note that the TTree in the last repro is really nothing special, it just has one integer branch with all values equal to 1

**Comment by pcanal** [comment]
Is the histogram of the same type? (TH1I vs TH1F)

**Comment by eguiraud** [comment]
in the snippet above, `h` is a `TH1F`

**Comment by pcanal** [comment]
Then it definitively weird :( ... Would have to see whether the cutoff is down in TSelectorDraw itself or somehow in the histogram ...

**Comment by Axel-Naumann** [comment]
Simple repo:
```C++
void repo() {
  int i = 10000;
  TTree t("t", "title");
  t.Branch("b", &i, "i/I");
  for (int e = 0; e < 50000000; ++e) {
    t.Fill();
  }
  t.Draw("b.i");
  std::cout << ((TH1*)gDirectory->FindObject("htemp"))->GetMaximum() << '\n';
}
```
Prints `1.67772e+07` and really shouldn't.

**Comment by stwunsch** [comment]
Here is a plain C++ reproducer, without ROOT ;)

```cpp
#include <iostream>

int main() {
    float x = 0.f;
    for (int i = 0u; i < 50000000; i++) {
        x += double(1.0);
    }
    std::cout << x << std::endl;
    // returns the magic number 1.67772e+07
}
```

The culprit in ROOT is precisely here: https://github.com/root-project/root/blob/master/tree/treeplayer/src/TSelectorDraw.cxx#L1361

**Comment by eguiraud** [comment]
Turns out this is a duplicate of https://sft.its.cern.ch/jira/browse/ROOT-2956

**Comment by Axel-Naumann** [comment]
@lmoneta given that we didn't understand this, and that this is likely not obvious to our users, could we add a test to `AddBinContent()` of the form:
```
if (w != 0 && bincontent + w == bincontent)
   Warning("AddBinContent", "Bin too full, adding %d is below its precision and does not change the bin content!", w);
```

**Comment by lmoneta** [comment]
Yes, but why using TH1F ?
Everybody should always use TH1D, unless there are some memory issues.
I have seen problem like this already too many times

**Comment by stwunsch** [comment]
From my experience, everyone is using TH1Fs because the datatype in most ntuple is float and not double (there because of the size). Most users may think that if the branch type is float, you should also use a TH1F. Not necessarily correct, but I can guarantee you that this happens a lot.

And more important, especially for NanoAOD, the following very typical user code will show a wrong result, on rather small files (see the initial reproducer):

```python
>>> import ROOT
>>> f = ROOT.TFile('DYJetsToLL.root')
>>> t = f.Get('Events')
>>> t.Draw('GenPart_pdgId')
```

The bare minimum would fixing the type of histogram, that `TTree::Draw` returns (TH1D, not TH1F), I guess.

**Comment by lmoneta** [comment]
This is completely wrong, an histogram type has nothing to do with the data type. The axis binning is always used in double precision.  We should probably then specify this better in the TH1 documentation!

**Comment by lmoneta** [comment]
And yes Three::Draw should returns TH1D. @pcanal, can this be changed ?
I guess RDataFrame.Histo1D returns TH1D already

**Comment by stwunsch** [comment]
> This is completely wrong, an histogram type has nothing to do with the data type. The axis binning is always used in double precision. We should probably then specify this better in the TH1 documentation!

And that's not how ROOT histograms are used ;)

**Comment by stwunsch** [comment]
![image](https://user-images.githubusercontent.com/6951222/110138875-b9476480-7dd2-11eb-997a-d2a636e5be00.png)

https://root.cern/doc/master/classTH1.html#associated-errors

I also would have understood the docs as the precision on the value that you fill in rather than the precision on the bin content. And it's also not a precision in that sense, it's a maximum bin content you can achieve by doing small increments. Actually, in principle you could have larger bin contents with float, e.g., with large weights.

**Comment by pcanal** [comment]
> And yes TTree::Draw should return a TH1D. @pcanal, can this be changed ?

This is straightforward to change (in TSelectorDraw), it is the obvious solution to the problem (albeit it actually just delay the problem per se .. but by a lot I guess :) ) but is not without visible consequences:

* Increase memory usage (would be visible by users that use a lot of Draws).
* Potential type error for those that call Draw and then retrieve the htmp histogram (lucky users would have provided their own .. .likely also a TH1F though !? ... the unlikely one would have C-style casted it to TH1F (which actually might still appears to work depending which routines they call on that pointer) ... and the most unlucky would have use dynamic_cast but forgot to test the result .... and the weird ones would rely on the saturation to detect some type of results (see [xkcd space bar heating](https://xkcd.com/1172/))

**Comment by Axel-Naumann** [comment]
Forget it: 🦔 

* https://savannah.web.cern.ch/HEP_Applications/savroot/bugs/63873.html
* https://savannah.web.cern.ch/HEP_Applications/savroot/bugs/85436.html
* Email conversation between Philippe, Olivier, Enrico (!), me, and rootdev, titled "TTree::Draw default histogram", in 2017, where we concluded "why not"
* We have c9e40d125fc3b27db05ced052856b38a814231d1
* I then changed the default to `TH1D`: 0ad637e2105e0ef248029f6377e5e73323333e00
* The main issue was user code that static-casts `htemp` to `TH1F` : if they then access the `TArrayF` (that's actually a `TArrayD`) then that's very bad. We even had such user code in ROOT: a9570d37277a13f7871ee64e0f522725fc250e1e
* So in d001610754d0d620faac7400027086ea73063b74 I turned the default back to `TH1F`

Hey, what about https://github.com/root-project/root/issues/6671#issuecomment-791443564 ?

**Comment by stwunsch** [comment]
What's the performance impact of https://github.com/root-project/root/issues/6671#issuecomment-791443564? I guess that's the question. Otherwise, I'd be all in!

**Comment by ferdymercury** [comment]
> What's the performance impact of [#6671 (comment)](https://github.com/root-project/root/issues/6671#issuecomment-791443564)? I guess that's the question. Otherwise, I'd be all in!

A warning would be helpful indeed, but maybe only for the first time this happens? Just to avoid getting 1000 lines printed out in the terminal.

Just as a side note, there is another TTree.Draw bug out there concerning double (or ULong) precision / failing silently as in this issue:
https://sft.its.cern.ch/jira/browse/ROOT-8009

**Comment by Axel-Naumann** [comment]
> A warning would be helpful indeed, but maybe only for the first time this happens?

Certainly - but the perf impact we're concerned about is the `if` in the case where no diagnostic is emitted, not the subsequent warning.

**Comment by Axel-Naumann** [comment]
Removing v6.24 as milestone:
while I'd be happy to have this addressed in v6.24, this has a long, long history; we cannot claim it's suddenly a blocker after 25 years ;-)

**Comment by ferdymercury** [comment]
> Certainly - but the perf impact we're concerned about is the `if` in the case where no diagnostic is emitted, not the subsequent warning.

Ok, I see. Probably it doesn't make a difference performance-wise, but there are functions to calculate what the next available float value is:
`https://en.cppreference.com/w/cpp/numeric/math/nextafter`

Side note: we are already losing some performance in TH1I by checking for overflow, but not for TH1F.

See:
```
 void TH1I::AddBinContent(Int_t bin, Double_t w)
 {
    Long64_t newval = fArray[bin] + Long64_t(w);
    if (newval > -INT_MAX && newval < INT_MAX) {fArray[bin] = Int_t(newval); return;}
    if (newval < -INT_MAX) fArray[bin] = -INT_MAX;
    if (newval >  INT_MAX) fArray[bin] =  INT_MAX;
 }
```

vs

```
void TH1F::AddBinContent(Int_t bin, Double_t w)
                                 {fArray[bin] += Float_t (w);}
```

**Comment by dpiparo** [comment]
I'd like to wrap this up. Maybe I am naive but, as pointed out by Rene on [JIRA](https://sft.its.cern.ch/jira/browse/ROOT-2956), isn't this understood? We are trying to increment by one a single precision floating point number and at some point, the gap between 2 subsequent 32bit fp numbers is larger than 1...

**Comment by eguiraud** [comment]
My two cents.

The root cause of the problem is understood and it is what you mentioned.

However, the way this presents to users is that if they run the following code on a large-enough inputs:

```python
>>> f = ROOT.TFile('DYJetsToLL.root')
>>> t = f.Get('Events')
>>> t.Draw('GenPart_pdgId')
```

users _silently_ get a completely wrong histogram:

![image](https://github.com/root-project/root/assets/10999034/1d53b401-cf13-471c-bf6f-908835845b20)

instead of the histogram they _should_ get:

![image](https://github.com/root-project/root/assets/10999034/4bfc1d5d-e7e6-450a-b3c0-783cfccbda3a)

Even if you are lucky enough that the histogram is so wrong that you spot the problem immediately, it is then very difficult to go from "my histogram is wrong" to "low floating point precision makes `Fill` effectively a no-op inside `TTree::Draw`". So difficult that I do not think one can reasonably expect users to diagnose the issue themselves.

With that said, new interfaces already moved to TH1D by default instead of TH1F, and TTree::Draw cannot be changed to produce a TH1D for the backward-compat issues Axel listed above, so the only remaining option, AFAICT, is to add a check after every `Fill` or deprecate `TTree::Draw`.

**Comment by guitargeek** [comment]
Thanks Enrico, I was just about to write the same :laughing: I think once we implement such a precision loss check in `TH*F::Fill()` we should be fine. I would be hesitant to implement this for general `TH` types because of performance overhead. But given that we encourage people to use `THD` because of the reasons that became apparent in this discussion, I think it's fine to do this for `THF`. What is you opinion on this? Deprecating `TTree::Draw()` is a whole different beast that is maybe outside the scope of this discussion

**Comment by lmoneta** [comment]
Hi, 
I agree to add the precision check in THF, and we already have this check-in THI, THS and THC!

**Comment by ferdymercury** [comment]
According to https://en.cppreference.com/w/cpp/numeric/math/nextafter,

TH1F stops working well at 1e7 (with integer weights). Should we add this as the maximum value for TH1F, as is with TH1C for 128 e.g.
With non-integer weights, this might become more of a problem to check as it is highly dependent on the chosen weight. But usually w = 1.

```
Precision loss demo for float:
nextafter(1e+01, INF) gives 10.000001; Δ = 0.000001
nextafter(1e+02, INF) gives 100.000008; Δ = 0.000008
nextafter(1e+03, INF) gives 1000.000061; Δ = 0.000061
nextafter(1e+04, INF) gives 10000.000977; Δ = 0.000977
nextafter(1e+05, INF) gives 100000.007812; Δ = 0.007812
nextafter(1e+06, INF) gives 1000000.062500; Δ = 0.062500
nextafter(1e+07, INF) gives 10000001.000000; Δ = 1.000000
nextafter(1e+08, INF) gives 100000008.000000; Δ = 8.000000
```

**Comment by ferdymercury** [comment]
> Yes, but why using TH1F ? Everybody should always use TH1D, unless there are some memory issues. I have seen problem like this already too many times

If we want to encourage that change, I think we should start by removing TH1F from all the doxygen examples in ROOT, which is I believe why many people still use TH1F.

If you run a grep, there are almost 2000 results. Most of them in the tutorials and test folders. Others in roofit and tmva.

**Comment by eguiraud** [comment]
> why using TH1F ? Everybody should always use TH1D,

TTree::Draw

**Comment by vepadulano** [comment]
A summary of the discussion at the linked PR:

* We cannot implement a precision loss check in TH*F classes as they are implemented currently, as it would effectively be a no-op and a waste of CPU cycles
* We can better document the interfaces highlighting the possibility to incur in undefined behaviour of the type seen in this issue
* The only way for a user to know that they are reaching the limits of their histogram is to check the amount of entries and compare it with the maximum integer bin content for the float channel as documented by the changes in the PR
* TH1F should probably never be used unless the user really knows what they are doing. Unfortunately it is the default histogram type used by `TTree::Draw`, rendering its deprecation very complicated
* To combat this we should probably remove bare TH1F usage from tutorials everywhere

**Comment by fwyzard** [comment]
> * We cannot implement a precision loss check in TH*F classes as they are implemented currently, as it would effectively be a no-op ... 

Why not ?

You can always check that (value in the bin after fill) - (value in the bin before fill) is reasonably close to the value that was added, and print a warning message otherwise.


> ... and a waste of CPU cycles

Ah, yes, it would definitely be slower !

**Comment by ferdymercury** [comment]
> You can always check that (value in the bin after fill) - (value in the bin before fill) is reasonably close to the value that was added, and print a warning message otherwise.

Not really. Your suggestion would work well if you only had `AddBinContentByOne`. But if you have `AddBinContentByWeight`, then what's "close" becomes non-trivial. In other words, Closeness is a function of Weight. So your limit would depend on Weight. There is no way to ensure that that a user always calls AddBinContentByWeight with the same weight. There is no way to ensure that the user calls uses the same weight for each bin of the histogram.

This would result in different "overflow bin limits" for every bin in the histogram. So it's an ill-posed problem.

To me, the only solution is using TH1L where the overflow limit is well defined, and forget about floating precision.

**Comment by fwyzard** [comment]
Sorry, but I strongly disagree.

`TH1F` implements `Fill(x, w)` via `AddBinContent(bin, w)`:
```c++
void AddBinContent(Int_t bin, Double_t w) override
{
    fArray[bin] += Float_t (w);
}
```

If one wants to be warned about overflows, it could be changed to
```c++
void AddBinContent(Int_t bin, Double_t w) override
{
    auto old = fArray[bin];
    fArray[bin] += Float_t (w);
    auto inc = fArray[bin] - old;
    if (inc != w) {  // could be done with a non-exact comparison with some tolerance
      std::cerr << "Warning: TH1F::Fill(...) failed to increment the bin due to limited floating point precision\n";
    }
}
```

**Comment by ferdymercury** [comment]
> // could be done with a non-exact comparison with some tolerance

Yeah, that's what I meant. Please define a tolerance that scales over order of magnitudes and weights.

**Comment by fwyzard** [comment]
Sorry, I assumed that would be your job ?

**Comment by vepadulano** [comment]
Re-opening the issue following further discussion. The linked PR is still valid as it documents the current state of the implementation, so that doesn't need to be changed. An investigation into finding a tolerance that can account for different (orders of magnitude of) weights is the next step for this issue. Since it was not foreseen in the PoW for 2024, we cannot give an ETA at this moment.

**Comment by ferdymercury** [comment]
As alternative ideas:
- a new function SafeFill(x, w, tol) where the user can define his tolerance tol.
- a static function IsInOverflow(bin, w) that checks whether a histogram bin is in overflow depending on its content and a potential weight w to be added.

From my point of view, I will just go towards TH1D or TH1L and away from TTree::Draw

**Comment by fwyzard** [comment]
Here is a implementation that may be naive, but I would argue catches the vast majority of the use cases:
```c++
constexpr bool compare(float expected, float actual) {
  // most simple and most common case
  if (actual == expected)
    return true;

  // comparison with an arbitrary small tolerance 
  constexpr const float epsilon = std::numeric_limits<float>::epsilon();
  constexpr const float delta = std::fabs(expected) * epsilon;
  if (actual > expected - delta) and (actual < expected + delta)
    return true;
  
  return false;
} 
```

If any of the arguments (the weight or the actual increment) is NaN or infinite the function should return `false`, which kind of makes sense in [the above context](https://github.com/root-project/root/issues/6671#issuecomment-2047035457).

**Comment by ferdymercury** [comment]
With @lmoneta we were discussing in the PR this kind of case:

a histogram with an initial SetBinContent of 1e8, and you add an event with weight 8. This leads to an error

`(1e8f+8.01f)- 1e8f  - 8.01f = -0.01f`

which compared to the bin content of 1e8 is a negligible difference.

But `compare(8.01f,8.00f)` would return that the increment is not the same.

So we were thinking of defining somehow a relative tolerance. We used std::nextafterf and compared the relative distance wrt the original, and divided by w.
But weird things may happen here, because you might call Fill with a negative weight, and the result might come close to zero for some bins, so a relative normalization is also ugly. We would need a compromise somehow between an absolute and a relative normalization for the tolerance, or adding a lot of CPU-wasting checks. Or just focus on the main cases with positive weights.

**Comment by fwyzard** [comment]
Relative with respect to the bin value (before the increment), or with respect to the increment, or with respect to the "correct" bin value after the increment ?

**Comment by dpiparo** [comment]
old histograms will not receive any update
https://root.cern/doc/v640/release-notes.html#histograms
